bundles/org.openhab.binding.freeathomesystem/src/main/java/org/openhab/binding/freeathomesystem/internal/handler/FreeAtHomeBridgeHandler.java

```diff
@@ -0,0 +1,832 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.freeathomesystem.internal.handler;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.AuthenticationStore;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.util.BasicAuthentication;
+import org.eclipse.jetty.client.util.StringContentProvider;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.eclipse.jetty.util.thread.QueuedThreadPool;
+import org.eclipse.jetty.websocket.client.ClientUpgradeRequest;
+import org.eclipse.jetty.websocket.client.WebSocketClient;
+import org.openhab.binding.freeathomesystem.internal.FreeAtHomeSystemDiscoveryService;
+import org.openhab.binding.freeathomesystem.internal.configuration.FreeAtHomeBridgeHandlerConfiguration;
+import org.openhab.binding.freeathomesystem.internal.datamodel.FreeAtHomeDeviceDescription;
+import org.openhab.binding.freeathomesystem.internal.util.FreeAtHomeHttpCommunicationException;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseBridgeHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonParser;
+import com.google.gson.stream.JsonReader;
+
+/**
+ * The {@link FreeAtHomeBridgeHandler} is responsible for handling the free@home bridge and
+ * its main communication.
+ *
+ * @author Andras Uhrin - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class FreeAtHomeBridgeHandler extends BaseBridgeHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(FreeAtHomeBridgeHandler.class);
+
+    public ChannelUpdateHandler channelUpdateHandler = new ChannelUpdateHandler();
```

Comment:
The way this is exposed and used by multiple handler objects might be dangerous and not thread-safe. I didn't fully analyze this yet, so will need to get back to this.

Reply:
Okay, after having another look - it won't handle concurrency well. It looks like an implementation of an observer pattern where the child handlers can subscribe to events from the bridge handler.

With some small adjustments this could be made more robust. Instead of sharing the object directly with the thing handlers, you could manage the subscriptions internally within the bridge handler. A quite straight-forward and similar approach can be found in the Miele@home binding.

In the child handler's `initialize` method:

https://github.com/openhab/openhab-addons/blob/3d1c6baf493e7ca834a9448c7c29f335f07b2c35/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleApplianceHandler.java#L163-L166

Please note that the child handler implements this interface:

https://github.com/openhab/openhab-addons/blob/3d1c6baf493e7ca834a9448c7c29f335f07b2c35/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/ApplianceStatusListener.java#L28-L55

The bridge handler simply keeps track of listeners/subscribers using a `Map`:

https://github.com/openhab/openhab-addons/blob/3d1c6baf493e7ca834a9448c7c29f335f07b2c35/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleBridgeHandler.java#L107

managed here:

https://github.com/openhab/openhab-addons/blob/3d1c6baf493e7ca834a9448c7c29f335f07b2c35/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleBridgeHandler.java#L503-L540